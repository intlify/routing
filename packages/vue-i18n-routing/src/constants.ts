/** @internal */
export const STRATEGIES = {
  PREFIX: 'prefix',
  PREFIX_EXCEPT_DEFAULT: 'prefix_except_default',
  PREFIX_AND_DEFAULT: 'prefix_and_default',
  NO_PREFIX: 'no_prefix'
} as const

// NOTE:
//  we avoid SSR issue with the temp variables
//  (I think it seem to not be able to handle the MemberExpression case)
// - https://github.com/vitejs/vite/pull/6171
// - https://github.com/vitejs/vite/pull/3848
/*
export const VUE_I18N_ROUTING_DEFAULTS = {
  defaultLocale: '',
  strategy: 'prefix_except_default',
  trailingSlash: false,
  routesNameSeparator: '___',
  defaultLocaleRouteNameSuffix: 'default'
}
*/

/** @internal */
export const DEFAULT_LOCALE = ''
/** @internal */
export const DEFAULT_STRATEGY = STRATEGIES.PREFIX_EXCEPT_DEFAULT
/** @internal */
export const DEFAULT_TRAILING_SLASH = false
/** @internal */
export const DEFAULT_ROUTES_NAME_SEPARATOR = '___'
/** @internal */
export const DEFAULT_LOCALE_ROUTE_NAME_SUFFIX = 'default'
/** @internal */
export const DEFAULT_DETECTION_DIRECTION = 'ltr'
/** @internal */
export const DEFAULT_BASE_URL = ''
/** @internal */
export const DEFAULT_DYNAMIC_PARAMS_KEY = ''
