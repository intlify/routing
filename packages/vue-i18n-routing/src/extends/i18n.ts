import { isBoolean, isObject } from '@intlify/shared'
import { ref, computed, watch, isVue3, effectScope, isVue2 } from 'vue-demi'

import {
  localePath,
  localeRoute,
  localeLocation,
  switchLocalePath,
  getRouteBaseName,
  resolveRoute,
  localeHead
} from '../compatibles'
import { DEFAULT_BASE_URL } from '../constants'
import { resolveBaseUrl, isVueI18n, getComposer, inBrowser } from '../utils'

import type { I18nRoutingOptions, LocaleObject } from '../types'
import type { I18n, Composer, VueI18n } from '@intlify/vue-i18n-bridge'
import type { App } from 'vue-demi'

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type Vue = any

// eslint-disable-next-line @typescript-eslint/ban-types
export function proxyVueInstance(target: Function): Function {
  // `this` is the Vue instance
  return function (this: Vue) {
    return Reflect.apply(
      target,
      {
        getRouteBaseName: this.getRouteBaseName,
        localePath: this.localePath,
        localeRoute: this.localeRoute,
        localeLocation: this.localeLocation,
        resolveRoute: this.resolveRoute,
        switchLocalePath: this.switchLocalePath,
        localeHead: this.localeHead,
        i18n: this.$i18n,
        route: this.$route,
        router: this.$router
      },
      // eslint-disable-next-line prefer-rest-params
      arguments
    )
  }
}

/**
 * An options of Vue I18n Routing Plugin
 */
export interface VueI18nRoutingPluginOptions {
  /**
   * Whether to inject some option APIs style methods into Vue instance
   *
   * @defaultValue `true`
   */
  inject?: boolean
  /**
   * @internal
   */
  __composerExtend?: (composer: Composer) => void
  /**
   * @internal
   */
  __vueI18nExtend?: (vueI18n: VueI18n) => void
}

export interface ExtendProperyDescripters {
  [key: string]: Pick<PropertyDescriptor, 'get'>
}
export type ExtendComposerHook = (compser: Composer) => void
export type ExtendVueI18nHook = (composer: Composer) => ExtendProperyDescripters
export type ExtendExportedGlobalHook = (global: Composer) => ExtendProperyDescripters

export interface ExtendHooks {
  onExtendComposer?: ExtendComposerHook
  onExtendExportedGlobal?: ExtendExportedGlobalHook
  onExtendVueI18n?: ExtendVueI18nHook
}

export type VueI18nExtendOptions<Context = unknown> = Pick<I18nRoutingOptions<Context>, 'baseUrl'> & {
  locales?: string[] | LocaleObject[]
  localeCodes?: string[]
  context?: Context
  hooks?: ExtendHooks
}

export function extendI18n<Context = unknown, TI18n extends I18n = I18n>(
  i18n: TI18n,
  {
    locales = [],
    localeCodes = [],
    baseUrl = DEFAULT_BASE_URL,
    hooks = {},
    context = {} as Context
  }: VueI18nExtendOptions<Context> = {}
) {
  const scope = effectScope()

  const orgInstall = i18n.install
  i18n.install = (vue: Vue, ...options: unknown[]) => {
    const pluginOptions = isPluginOptions(options[0]) ? options[0] : { inject: true }
    if (pluginOptions.inject == null) {
      pluginOptions.inject = true
    }
    pluginOptions.__composerExtend = (c: Composer) => {
      const g = getComposer(i18n)
      c.locales = computed(() => g.locales.value)
      c.localeCodes = computed(() => g.localeCodes.value)
      c.baseUrl = computed(() => g.baseUrl.value)
    }
    if (isVueI18n(i18n.global)) {
      pluginOptions.__vueI18nExtend = (vueI18n: VueI18n) => {
        extendVueI18n(vueI18n, hooks.onExtendVueI18n)
      }
    }

    Reflect.apply(orgInstall, i18n, [vue, ...options])

    const composer = getComposer(i18n)

    // extend global
    scope.run(() => extendComposer(composer, { locales, localeCodes, baseUrl, hooks, context }))
    if (isVueI18n(i18n.global)) {
      extendVueI18n(i18n.global, hooks.onExtendVueI18n)
    }

    // extend vue component instance for Vue 3
    const app = vue as App
    // prettier-ignore
    const exported = i18n.mode === 'composition'
      ? isVue3
        ? app.config.globalProperties.$i18n
        : i18n
      : isVue2
        ? i18n
        : null
    if (exported) {
      extendExportedGlobal(exported, composer, hooks.onExtendExportedGlobal)
    }

    if (pluginOptions.inject) {
      // extend vue component instance
      vue.mixin({
        methods: {
          resolveRoute: proxyVueInstance(resolveRoute),
          localePath: proxyVueInstance(localePath),
          localeRoute: proxyVueInstance(localeRoute),
          localeLocation: proxyVueInstance(localeLocation),
          switchLocalePath: proxyVueInstance(switchLocalePath),
          getRouteBaseName: proxyVueInstance(getRouteBaseName),
          localeHead: proxyVueInstance(localeHead)
        }
      })
    }

    // release scope on unmounting
    if (app.unmount) {
      const unmountApp = app.unmount
      app.unmount = () => {
        scope.stop()
        unmountApp()
      }
    }
  }

  return scope
}

function extendComposer<Context = unknown>(composer: Composer, options: VueI18nExtendOptions<Context>) {
  const { locales, localeCodes, baseUrl, context } = options

  const _locales = ref<string[] | LocaleObject[]>(locales!)
  const _localeCodes = ref<string[]>(localeCodes!)
  const _baseUrl = ref<string>('')

  composer.locales = computed(() => _locales.value)
  composer.localeCodes = computed(() => _localeCodes.value)
  composer.baseUrl = computed(() => _baseUrl.value)

  if (inBrowser) {
    watch(
      composer.locale,
      () => {
        _baseUrl.value = resolveBaseUrl(baseUrl!, context!)
      },
      { immediate: true }
    )
  } else {
    _baseUrl.value = resolveBaseUrl(baseUrl!, context!)
  }

  if (options.hooks && options.hooks.onExtendComposer) {
    options.hooks.onExtendComposer(composer)
  }
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function extendExportedGlobal(exported: any, g: Composer, hook?: ExtendExportedGlobalHook) {
  const properties: ExtendProperyDescripters[] = [
    {
      locales: {
        get() {
          return g.locales.value
        }
      },
      localeCodes: {
        get() {
          return g.localeCodes.value
        }
      },
      baseUrl: {
        get() {
          return g.baseUrl.value
        }
      }
    }
  ]
  hook && properties.push(hook(g))
  for (const property of properties) {
    for (const [key, descriptor] of Object.entries(property)) {
      Object.defineProperty(exported, key, descriptor)
    }
  }
}

function extendVueI18n(vueI18n: VueI18n, hook?: ExtendVueI18nHook): void {
  const composer = getComposer(vueI18n)
  const properties: ExtendProperyDescripters[] = [
    {
      locales: {
        get() {
          return composer.locales.value
        }
      },
      localeCodes: {
        get() {
          return composer.localeCodes.value
        }
      },
      baseUrl: {
        get() {
          return composer.baseUrl.value
        }
      }
    }
  ]
  hook && properties.push(hook(composer))
  for (const property of properties) {
    for (const [key, descriptor] of Object.entries(property)) {
      Object.defineProperty(vueI18n, key, descriptor)
    }
  }
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function isPluginOptions(options: any): options is VueI18nRoutingPluginOptions {
  return isObject(options) && 'inject' in options && isBoolean(options.inject)
}
