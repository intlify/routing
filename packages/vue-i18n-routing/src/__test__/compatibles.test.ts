/* eslint-disable @typescript-eslint/no-explicit-any */

import { makeSymbol } from '@intlify/shared'
import { createI18n } from '@intlify/vue-i18n-bridge'
import { createMemoryHistory } from '@intlify/vue-router-bridge'
import { describe, it, assert, expect } from 'vitest'

import { useSetup } from '../../scripts/vitest'
import { STRATEGIES } from '../constants'
import { createRouter } from '../extends/router'

import type { I18n } from '@intlify/vue-i18n-bridge'

describe('getRouteBaseName', () => {
  describe('vue-i18n composition mode', () => {
    it('should return base name', async () => {
      const i18n = createI18n({ legacy: false, locale: 'en' })
      const router = createRouter(i18n, {
        version: 4,
        locales: ['en', 'ja'],
        routes: [
          {
            path: '/',
            component: {
              template: '<div>Home</div>'
            },
            name: 'home'
          }
        ],
        history: createMemoryHistory()
      })
      await router.push('/en')
      const vm = useSetup(() => {}, [[router], [i18n]]) as any // FIXME:
      const name = vm.getRouteBaseName()
      assert.equal(name, 'home')
    })
  })

  describe('vue-i18n legacy mode', () => {
    it('should return base name', async () => {
      const i18n = createI18n({ locale: 'en' })
      const router = createRouter(i18n, {
        version: 4,
        locales: ['en', 'ja'],
        routes: [
          {
            path: '/',
            component: {
              template: '<div>Home</div>'
            },
            name: 'home'
          }
        ],
        history: createMemoryHistory()
      })
      await router.push('/en')
      const vm = useSetup(() => {}, [[router], [i18n]]) as any // FIXME:
      const name = vm.getRouteBaseName()
      assert.equal(name, 'home')
    })
  })
})

describe('localePath', () => {
  function testLocalePath(legacy = false) {
    ;[STRATEGIES.PREFIX_EXCEPT_DEFAULT, STRATEGIES.PREFIX_AND_DEFAULT, STRATEGIES.PREFIX].forEach(strategy => {
      describe(`route strategy: ${strategy}`, () => {
        it('should be worked', async () => {
          const warn = vi.spyOn(console, 'warn').mockImplementation(() => {})

          const i18n = createI18n({ legacy, locale: 'en' })
          const router = createRouter(i18n, {
            version: 4,
            locales: ['en', 'ja'],
            strategy,
            routes: [
              { path: '/', name: 'index', component: { template: '<div>index</div>' } },
              { path: '/about', name: 'about', component: { template: '<div>About</div>' } },
              { path: '/:pathMatch(.*)*', name: 'not-found', component: { template: '<div>Not Found</div>' } }
            ],
            history: createMemoryHistory()
          })
          await router.push('/en')
          const vm = useSetup(() => {}, [[router], [i18n]]) as any // FIXME:

          // path
          assert.equal(vm.localePath('/'), '/en')
          assert.equal(vm.localePath('/about', 'ja'), '/ja/about')
          // name
          assert.equal(vm.localePath('index', 'ja'), '/ja')
          assert.equal(vm.localePath('about'), '/en/about')
          assert.equal(vm.localePath('not-found', 'ja'), '/ja')
          assert.equal(vm.localePath('not-found'), '/en')
          // object
          assert.equal(vm.localePath({ name: 'about' }, 'ja'), '/ja/about')
          // omit name & path
          assert.equal(vm.localePath({ state: { foo: 1 } }), '/en')
          await router.push('/ja')
          assert.equal(vm.localePath({ state: { foo: 1 } }), '/ja')

          // preserve query parameters
          assert.equal(vm.localePath({ query: { foo: 1 } }), '/ja?foo=1')
          assert.equal(vm.localePath({ path: '/', query: { foo: 1 } }), '/ja?foo=1')
          assert.equal(vm.localePath({ name: 'about', query: { foo: 1 } }), '/ja/about?foo=1')
          assert.equal(vm.localePath({ path: '/about', query: { foo: 1 } }), '/ja/about?foo=1')
          assert.equal(vm.localePath('/?foo=1'), '/ja?foo=1')
          assert.equal(vm.localePath('/about?foo=1'), '/ja/about?foo=1')
          assert.equal(vm.localePath('/about?foo=1&test=2'), '/ja/about?foo=1&test=2')

          // no define path
          assert.equal(vm.localePath('/vue-i18n'), '/ja/vue-i18n')
          // no define name
          assert.equal(vm.localePath('vue-i18n'), '')

          // for vue-router deprecation
          // https://github.com/vuejs/router/blob/main/packages/router/CHANGELOG.md#414-2022-08-22
          expect(warn).not.toHaveBeenCalledWith('Discarded invalid param(s)')
        })
      })
    })

    describe(`route strategy: ${STRATEGIES.NO_PREFIX}`, () => {
      it('should be worked', async () => {
        const warn = vi.spyOn(console, 'warn').mockImplementation(() => {})

        const i18n = createI18n({ legacy, locale: 'en' })
        const router = createRouter(i18n, {
          version: 4,
          locales: ['en', 'ja'],
          strategy: 'no_prefix',
          routes: [
            { path: '/', name: 'index', component: { template: '<div>index</div>' } },
            { path: '/about', name: 'about', component: { template: '<div>About</div>' } },
            { path: '/:pathMatch(.*)*', name: 'not-found', component: { template: '<div>Not Found</div>' } }
          ],
          history: createMemoryHistory()
        })
        await router.push('/')
        const vm = useSetup(() => {}, [[router], [i18n]]) as any // FIXME:

        // path
        assert.equal(vm.localePath('/'), '/')
        assert.equal(vm.localePath('/about', 'ja'), '/about')
        // name
        assert.equal(vm.localePath('index', 'ja'), '/')
        assert.equal(vm.localePath('about'), '/about')
        assert.equal(vm.localePath('not-found', 'ja'), '/')
        assert.equal(vm.localePath('not-found'), '/')
        // object
        assert.equal(vm.localePath({ name: 'about' }, 'ja'), '/about')
        // omit name & path
        assert.equal(vm.localePath({ state: { foo: 1 } }), '/')
        await router.push('/ja')
        assert.equal(vm.localePath({ state: { foo: 1 } }), '/')
        // no define path
        assert.equal(vm.localePath('/vue-i18n'), '/vue-i18n')
        // no define
        assert.equal(vm.localePath('vue-i18n'), '')

        // for vue-router deprecation
        // https://github.com/vuejs/router/blob/main/packages/router/CHANGELOG.md#414-2022-08-22
        expect(warn).not.toHaveBeenCalledWith('Discarded invalid param(s)')
      })
    })
  }

  describe('composition mode', () => {
    testLocalePath()
  })

  describe('legacy mode', () => {
    testLocalePath(false)
  })
})

describe('localeRoute', () => {
  function testLocaleRoute(i18n: I18n) {
    it('should return route', async () => {
      const warn = vi.spyOn(console, 'warn').mockImplementation(() => {})

      const router = createRouter(i18n, {
        version: 4,
        locales: ['en', 'ja'],
        routes: [
          { path: '/', name: 'index', component: { template: '<div>index</div>' } },
          { path: '/about', name: 'about', component: { template: '<div>About</div>' } },
          { path: '/:pathMatch(.*)*', name: 'not-found', component: { template: '<div>Not Found</div>' } }
        ],
        history: createMemoryHistory()
      })
      await router.push('/en')
      const vm = useSetup(() => {}, [[router], [i18n]]) as any // FIXME:

      // path
      assert.include(vm.localeRoute('/'), {
        fullPath: '/en',
        path: '/en',
        // name: 'not-found___en',
        name: 'index___en',
        href: '/en'
      })
      assert.include(vm.localeRoute('/about', 'ja'), {
        fullPath: '/ja/about',
        path: '/ja/about',
        name: 'about___ja',
        href: '/ja/about'
      })
      assert.include(vm.localeRoute('/:pathMatch(.*)*', 'ja'), {
        fullPath: '/ja/:pathMatch(.*)*',
        path: '/ja/:pathMatch(.*)*',
        name: 'not-found___ja',
        href: '/ja/:pathMatch(.*)*'
      })
      // name
      assert.include(vm.localeRoute('index', 'ja'), {
        fullPath: '/ja',
        path: '/ja',
        name: 'index___ja',
        href: '/ja'
      })
      assert.include(vm.localeRoute('about'), {
        fullPath: '/en/about',
        path: '/en/about',
        name: 'about___en',
        href: '/en/about'
      })
      assert.include(vm.localeRoute('not-found', 'ja'), {
        fullPath: '/ja',
        path: '/ja',
        name: 'not-found___ja',
        href: '/ja'
      })
      assert.include(vm.localeRoute('not-found'), {
        fullPath: '/en',
        path: '/en',
        name: 'not-found___en',
        href: '/en'
      })
      // object
      assert.include(vm.localeRoute({ name: 'about' }, 'ja'), {
        fullPath: '/ja/about',
        path: '/ja/about',
        name: 'about___ja',
        href: '/ja/about'
      })
      // no define path
      assert.include(vm.localeRoute('/vue-i18n', 'ja'), {
        fullPath: '/ja/vue-i18n',
        path: '/ja/vue-i18n',
        name: 'not-found___ja',
        href: '/ja/vue-i18n'
      })
      // no define name
      assert.isUndefined(vm.localeRoute('vue-i18n'))

      // for vue-router deprecation
      // https://github.com/vuejs/router/blob/main/packages/router/CHANGELOG.md#414-2022-08-22
      expect(warn).not.toHaveBeenCalledWith('Discarded invalid param(s)')
    })
  }

  describe('composition mode', () => {
    const i18n = createI18n({ legacy: false, locale: 'en' })
    testLocaleRoute(i18n)
  })

  describe('legacy mode', () => {
    const i18n = createI18n({ locale: 'en' })
    testLocaleRoute(i18n)
  })
})

describe('localeLocation', () => {
  function testLocaleLocation(i18n: I18n) {
    it('should return route', async () => {
      const warn = vi.spyOn(console, 'warn').mockImplementation(() => {})

      const router = createRouter(i18n, {
        version: 4,
        locales: ['en', 'ja'],
        routes: [
          { path: '/', name: 'index', component: { template: '<div>index</div>' } },
          { path: '/about', name: 'about', component: { template: '<div>About</div>' } },
          { path: '/:pathMatch(.*)*', name: 'not-found', component: { template: '<div>Not Found</div>' } }
        ],
        history: createMemoryHistory()
      })
      await router.push('/en')
      const vm = useSetup(() => {}, [[router], [i18n]]) as any // FIXME:

      // path
      assert.include(vm.localeLocation('/'), {
        fullPath: '/en',
        path: '/en',
        // name: 'not-found___en',
        name: 'index___en',
        href: '/en'
      })
      assert.include(vm.localeLocation('/about', 'ja'), {
        fullPath: '/ja/about',
        path: '/ja/about',
        name: 'about___ja',
        href: '/ja/about'
      })
      assert.include(vm.localeLocation('/:pathMatch(.*)*', 'ja'), {
        fullPath: '/ja/:pathMatch(.*)*',
        path: '/ja/:pathMatch(.*)*',
        name: 'not-found___ja',
        href: '/ja/:pathMatch(.*)*'
      })
      // name
      assert.include(vm.localeLocation('index', 'ja'), {
        fullPath: '/ja',
        path: '/ja',
        name: 'index___ja',
        href: '/ja'
      })
      assert.include(vm.localeLocation('about'), {
        fullPath: '/en/about',
        path: '/en/about',
        name: 'about___en',
        href: '/en/about'
      })
      assert.include(vm.localeLocation('not-found', 'ja'), {
        fullPath: '/ja',
        path: '/ja',
        name: 'not-found___ja',
        href: '/ja'
      })
      assert.include(vm.localeLocation('not-found'), {
        fullPath: '/en',
        path: '/en',
        name: 'not-found___en',
        href: '/en'
      })
      // object
      assert.include(vm.localeLocation({ name: 'about' }, 'ja'), {
        fullPath: '/ja/about',
        path: '/ja/about',
        name: 'about___ja',
        href: '/ja/about'
      })
      // no define path
      assert.include(vm.localeLocation('/vue-i18n', 'ja'), {
        fullPath: '/ja/vue-i18n',
        path: '/ja/vue-i18n',
        name: 'not-found___ja',
        href: '/ja/vue-i18n'
      })
      // no define name
      assert.isUndefined(vm.localeLocation('vue-i18n'))

      // for vue-router deprecation
      // https://github.com/vuejs/router/blob/main/packages/router/CHANGELOG.md#414-2022-08-22
      expect(warn).not.toHaveBeenCalledWith('Discarded invalid param(s)')
    })
  }

  describe('composition mode', () => {
    const i18n = createI18n({ legacy: false, locale: 'en', globalInjection: true })
    testLocaleLocation(i18n)
  })

  describe('legacy mode', () => {
    const i18n = createI18n({ locale: 'en' })
    testLocaleLocation(i18n)
  })
})

describe('switchLocalePath', () => {
  function testSwtichLocalePath(i18n: I18n) {
    it('should be worked', async () => {
      const warn = vi.spyOn(console, 'warn').mockImplementation(() => {})

      const key = makeSymbol('i18n')
      const router = createRouter(i18n, {
        version: 4,
        locales: ['en', 'ja', 'fr'],
        dynamicRouteParamsKey: key,
        routes: [
          { path: '/', name: 'index', component: { template: '<div>index</div>' } },
          { path: '/about', name: 'about', component: { template: '<div>About</div>' } },
          {
            path: '/category/:id',
            name: 'category',
            meta: {
              [key]: {
                en: { id: 'english' },
                ja: { id: 'japanese' },
                fr: { id: 'franch' }
              }
            },
            component: { template: '<div>Category</div>' }
          },
          {
            path: '/count/:id',
            name: 'count',
            component: { template: '<div>Category</div>' }
          },
          {
            path: '/:pathMatch(.*)*',
            name: 'not-found',
            meta: {
              [key]: {
                en: { pathMatch: 'not-found-english' },
                ja: { pathMatch: 'not-found-japanese' },
                fr: { pathMatch: 'not-found-franch' }
              }
            },
            component: { template: '<div>Not Found</div>' }
          }
        ],
        history: createMemoryHistory()
      })
      await router.push('/en/about')
      const vm = useSetup(() => {}, [[router], [i18n]]) as any // FIXME:
      await router.push('/ja')

      assert.equal(vm.switchLocalePath('ja'), '/ja')
      assert.equal(vm.switchLocalePath('en'), '/en')
      assert.equal(vm.switchLocalePath('fr'), '/fr')
      assert.equal(vm.switchLocalePath('vue-i18n'), '')

      await router.push('/ja/about')

      assert.equal(vm.switchLocalePath('ja'), '/ja/about')
      assert.equal(vm.switchLocalePath('en'), '/en/about')
      assert.equal(vm.switchLocalePath('fr'), '/fr/about')
      assert.equal(vm.switchLocalePath('vue-i18n'), '')

      await router.push('/ja/about?foo=1&test=2')
      assert.equal(vm.switchLocalePath('ja'), '/ja/about?foo=1&test=2')
      assert.equal(vm.switchLocalePath('fr'), '/fr/about?foo=1&test=2')
      assert.equal(vm.switchLocalePath('en'), '/en/about?foo=1&test=2')

      await router.push('/ja/about?foo=bär&four=四&foo=bar')
      assert.equal(vm.switchLocalePath('ja'), '/ja/about?foo=b%C3%A4r&foo=bar&four=%E5%9B%9B')
      assert.equal(vm.switchLocalePath('fr'), '/fr/about?foo=b%C3%A4r&foo=bar&four=%E5%9B%9B')
      assert.equal(vm.switchLocalePath('en'), '/en/about?foo=b%C3%A4r&foo=bar&four=%E5%9B%9B')

      await router.push('/ja/about?foo=bär&four=四')
      assert.equal(vm.switchLocalePath('ja'), '/ja/about?foo=b%C3%A4r&four=%E5%9B%9B')
      assert.equal(vm.switchLocalePath('fr'), '/fr/about?foo=b%C3%A4r&four=%E5%9B%9B')
      assert.equal(vm.switchLocalePath('en'), '/en/about?foo=b%C3%A4r&four=%E5%9B%9B')

      await router.push('/ja/category/1')
      assert.equal(vm.switchLocalePath('ja'), '/ja/category/japanese')
      assert.equal(vm.switchLocalePath('en'), '/en/category/english')
      assert.equal(vm.switchLocalePath('fr'), '/fr/category/franch')

      await router.push('/ja/count/三')
      assert.equal(vm.switchLocalePath('ja'), '/ja/count/%E4%B8%89')
      assert.equal(vm.switchLocalePath('en'), '/en/count/%E4%B8%89')
      assert.equal(vm.switchLocalePath('fr'), '/fr/count/%E4%B8%89')

      await router.push('/ja/count/三?foo=bär&four=四&foo=bar')
      assert.equal(vm.switchLocalePath('ja'), '/ja/count/%E4%B8%89?foo=b%C3%A4r&foo=bar&four=%E5%9B%9B')
      assert.equal(vm.switchLocalePath('fr'), '/fr/count/%E4%B8%89?foo=b%C3%A4r&foo=bar&four=%E5%9B%9B')
      assert.equal(vm.switchLocalePath('en'), '/en/count/%E4%B8%89?foo=b%C3%A4r&foo=bar&four=%E5%9B%9B')

      await router.push('/ja/foo')
      assert.equal(vm.switchLocalePath('ja'), '/ja/not-found-japanese')
      assert.equal(vm.switchLocalePath('en'), '/en/not-found-english')
      assert.equal(vm.switchLocalePath('fr'), '/fr/not-found-franch')

      // for vue-router deprecation
      // https://github.com/vuejs/router/blob/main/packages/router/CHANGELOG.md#414-2022-08-22
      expect(warn).not.toHaveBeenCalledWith('Discarded invalid param(s)')
    })
  }

  describe('composition mode', () => {
    const i18n = createI18n({ legacy: false, locale: 'en' })
    testSwtichLocalePath(i18n)
  })

  describe('legacy mode', () => {
    const i18n = createI18n({ locale: 'en' })
    testSwtichLocalePath(i18n)
  })
})

describe('localeHead', () => {
  function testLocaleHead(i18n: I18n) {
    it('should be worked', async () => {
      const warn = vi.spyOn(console, 'warn').mockImplementation(() => {})

      const router = createRouter(i18n, {
        version: 4,
        locales: [
          {
            code: 'en',
            iso: 'en-US',
            dir: 'ltr'
          },
          {
            code: 'ja',
            iso: 'ja-JP'
          }
        ],
        baseUrl: 'http://localhost:8080',
        routes: [
          { path: '/', name: 'index', component: { template: '<div>index</div>' } },
          { path: '/about', name: 'about', component: { template: '<div>About</div>' } },
          { path: '/:pathMatch(.*)*', name: 'not-found', component: { template: '<div>Not Found</div>' } }
        ],
        history: createMemoryHistory()
      })
      await router.push('/en/about')
      const vm = useSetup(() => {}, [[router], [i18n]]) as any // FIXME:

      let head = vm.localeHead({ addDirAttribute: true, addSeoAttributes: true })
      expect(head).toMatchSnapshot('en')
      assert.equal(head.htmlAttrs.lang, 'en-US')

      await router.push('/ja')
      head = vm.localeHead({ addDirAttribute: true, addSeoAttributes: true })
      expect(head).toMatchSnapshot('ja')
      assert.equal(head.htmlAttrs.lang, 'ja-JP')

      // for vue-router deprecation
      // https://github.com/vuejs/router/blob/main/packages/router/CHANGELOG.md#414-2022-08-22
      expect(warn).not.toHaveBeenCalledWith('Discarded invalid param(s)')
    })
  }

  describe('composition mode', () => {
    const i18n = createI18n({ legacy: false, locale: 'en' })
    testLocaleHead(i18n)
  })

  describe('legacy mode', () => {
    const i18n = createI18n({ locale: 'en' })
    testLocaleHead(i18n)
  })
})

/* eslint-enable @typescript-eslint/no-explicit-any */
