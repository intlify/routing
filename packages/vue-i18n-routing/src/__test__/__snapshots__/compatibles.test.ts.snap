// Vitest Snapshot v1

exports[`localeHead > composition mode > should be worked > en 1`] = `
{
  "htmlAttrs": {
    "dir": "ltr",
    "lang": "en-US",
  },
  "link": [
    {
      "hid": "i18n-alt-en",
      "href": "http://localhost:8080/en/about",
      "hreflang": "en",
      "rel": "alternate",
    },
    {
      "hid": "i18n-alt-en-US",
      "href": "http://localhost:8080/en/about",
      "hreflang": "en-US",
      "rel": "alternate",
    },
    {
      "hid": "i18n-alt-ja",
      "href": "http://localhost:8080/ja/about",
      "hreflang": "ja",
      "rel": "alternate",
    },
    {
      "hid": "i18n-alt-ja-JP",
      "href": "http://localhost:8080/ja/about",
      "hreflang": "ja-JP",
      "rel": "alternate",
    },
    {
      "hid": "i18n-can",
      "href": "http://localhost:8080/en/about",
      "rel": "canonical",
    },
  ],
  "meta": [
    {
      "content": "http://localhost:8080/en/about",
      "hid": "i18n-og-url",
      "property": "og:url",
    },
    {
      "content": "en_US",
      "hid": "i18n-og",
      "property": "og:locale",
    },
    {
      "content": "ja_JP",
      "hid": "i18n-og-alt-ja-JP",
      "property": "og:locale:alternate",
    },
  ],
}
`;

exports[`localeHead > composition mode > should be worked > ja 1`] = `
{
  "htmlAttrs": {
    "dir": "ltr",
    "lang": "ja-JP",
  },
  "link": [
    {
      "hid": "i18n-alt-en",
      "href": "http://localhost:8080/en",
      "hreflang": "en",
      "rel": "alternate",
    },
    {
      "hid": "i18n-alt-en-US",
      "href": "http://localhost:8080/en",
      "hreflang": "en-US",
      "rel": "alternate",
    },
    {
      "hid": "i18n-alt-ja",
      "href": "http://localhost:8080/ja",
      "hreflang": "ja",
      "rel": "alternate",
    },
    {
      "hid": "i18n-alt-ja-JP",
      "href": "http://localhost:8080/ja",
      "hreflang": "ja-JP",
      "rel": "alternate",
    },
    {
      "hid": "i18n-can",
      "href": "http://localhost:8080/ja",
      "rel": "canonical",
    },
  ],
  "meta": [
    {
      "content": "http://localhost:8080/ja",
      "hid": "i18n-og-url",
      "property": "og:url",
    },
    {
      "content": "ja_JP",
      "hid": "i18n-og",
      "property": "og:locale",
    },
    {
      "content": "en_US",
      "hid": "i18n-og-alt-en-US",
      "property": "og:locale:alternate",
    },
  ],
}
`;

exports[`localeHead > legacy mode > should be worked > en 1`] = `
{
  "htmlAttrs": {
    "dir": "ltr",
    "lang": "en-US",
  },
  "link": [
    {
      "hid": "i18n-alt-en",
      "href": "http://localhost:8080/en/about",
      "hreflang": "en",
      "rel": "alternate",
    },
    {
      "hid": "i18n-alt-en-US",
      "href": "http://localhost:8080/en/about",
      "hreflang": "en-US",
      "rel": "alternate",
    },
    {
      "hid": "i18n-alt-ja",
      "href": "http://localhost:8080/ja/about",
      "hreflang": "ja",
      "rel": "alternate",
    },
    {
      "hid": "i18n-alt-ja-JP",
      "href": "http://localhost:8080/ja/about",
      "hreflang": "ja-JP",
      "rel": "alternate",
    },
    {
      "hid": "i18n-can",
      "href": "http://localhost:8080/en/about",
      "rel": "canonical",
    },
  ],
  "meta": [
    {
      "content": "http://localhost:8080/en/about",
      "hid": "i18n-og-url",
      "property": "og:url",
    },
    {
      "content": "en_US",
      "hid": "i18n-og",
      "property": "og:locale",
    },
    {
      "content": "ja_JP",
      "hid": "i18n-og-alt-ja-JP",
      "property": "og:locale:alternate",
    },
  ],
}
`;

exports[`localeHead > legacy mode > should be worked > ja 1`] = `
{
  "htmlAttrs": {
    "dir": "ltr",
    "lang": "ja-JP",
  },
  "link": [
    {
      "hid": "i18n-alt-en",
      "href": "http://localhost:8080/en",
      "hreflang": "en",
      "rel": "alternate",
    },
    {
      "hid": "i18n-alt-en-US",
      "href": "http://localhost:8080/en",
      "hreflang": "en-US",
      "rel": "alternate",
    },
    {
      "hid": "i18n-alt-ja",
      "href": "http://localhost:8080/ja",
      "hreflang": "ja",
      "rel": "alternate",
    },
    {
      "hid": "i18n-alt-ja-JP",
      "href": "http://localhost:8080/ja",
      "hreflang": "ja-JP",
      "rel": "alternate",
    },
    {
      "hid": "i18n-can",
      "href": "http://localhost:8080/ja",
      "rel": "canonical",
    },
  ],
  "meta": [
    {
      "content": "http://localhost:8080/ja",
      "hid": "i18n-og-url",
      "property": "og:url",
    },
    {
      "content": "ja_JP",
      "hid": "i18n-og",
      "property": "og:locale",
    },
    {
      "content": "en_US",
      "hid": "i18n-og-alt-en-US",
      "property": "og:locale:alternate",
    },
  ],
}
`;
